src/Storages/System/StorageSystemZooKeeper.cpp

```diff
@@ -222,6 +222,7 @@ class SystemZooKeeperSource : public ISource
     ContextPtr context;
     ZooKeeperWithFaultInjection::Ptr zookeeper;
     bool started = false;
+    std::unordered_set<String> visited;
```

Comment:
This is the real fix. As generate() might be called more than once now, we need to keep this alive between calls to compare and not duplicate across blocks.